src/pyodide-py/pyodide/_base.py

```diff
@@ -43,41 +43,149 @@ def open_url(url: str) -> StringIO:
     return StringIO(req.response)
 
 
-def quiet_code(code: str) -> bool:
+class CodeRunner:
     """
-    Does the last nonwhitespace character of code is a semicolon?
+    A code runner to serve eval_code and eval_code_async.
 
-    This can be overridden to customize the way eval_code is silenced.
+    Parameters
+    ----------
+    ns
+        `locals()` or `globals()` context where to execute code.
+        This namespace is updated by the subsequent calls to `run()`.
 
     Examples
     --------
-    >>> quiet_code('x + 1')
-    False
-    >>> quiet_code('x + 1 ;')
-    True
-    >>> quiet_code('x + 1 # comment ;')
-    False
+    >>> CodeRunner().run("1+1")
+    2
+    >>> CodeRunner().run("1+1;")
+    >>> runner = CodeRunner()
+    >>> runner.run("x = 5")
+    >>> runner.run("x + 1")
+    6
     """
-    # largely inspired from IPython:
-    # https://github.com/ipython/ipython/blob/86d24741188b0cedd78ab080d498e775ed0e5272/IPython/core/displayhook.py#L84
-
-    codeio = StringIO(code)
-    tokens = list(tokenize.generate_tokens(codeio.readline))
-
-    for token in reversed(tokens):
-        if token.type in (
-            tokenize.ENDMARKER,
-            tokenize.NL,  # ignoring empty lines (\n\n)
-            tokenize.NEWLINE,
-            tokenize.COMMENT,
-        ):
-            continue
-        return (token.type == tokenize.OP) and (token.string == ";")
 
-    return False
-
-
-def eval_code(code: str, ns: Dict[str, Any]) -> None:
+    def __init__(self, ns: Dict[str, Any] = None):
```

Comment:
Type annotations need an `Optional` here, but this can be adressed in a follow up PR. mypy in CI says that everything is OK which is very suspicious.